awswrangler/neptune/client.py

```diff
@@ -0,0 +1,205 @@
+        except Exception as e:
+            c.close()
+            raise e
+
+    
+    def read_sparql(self, query, headers:Dict[str, Any] = None) -> Dict[str, Any]:
+        if headers is None:
+            headers = {}
+        
+        data = {'query': query}
+
+        if 'content-type' not in headers:
+            headers['content-type'] = 'application/x-www-form-urlencoded'
+
+        uri = f'{self._http_protocol}://{self.host}:{self.port}/sparql'
+        req = self._prepare_request('POST', uri, data=data, headers=headers)
+        res = self._http_session.send(req)
+        return res
+
+
+    def status(self):
+        url = f'{self._http_protocol}://{self.host}:{self.port}/status'
+        req = self._prepare_request('GET', url, data='')
+        res = self._http_session.send(req)
+        if res.status_code == 200:
+            return res.json()
+        else:
+            _logger.error("Error connecting to Amazon Neptune cluster. Please verify your connection details")
+            raise ConnectionError(res.status_code)
+
+
+    def _gremlin_results_to_dict(self, result) -> Dict[str, Any]:
+        res=[]
+
+        # For lists or paths unwind them
+        if isinstance(result, list) or isinstance(result, Path):
+            for x in result:
+                res.append(self._parse_dict(x))
+
+        # For dictionaries just add them
+        elif isinstance(result, dict):
+            res.append(result)
+
+        # For everything else parse them
+        else:
+            res.append(self._parse_dict(result))            
+        return res
+
+
+    def _parse_dict(self, data) -> Dict[str, Any]:
+        d = dict()
+
+        # If this is a list or Path then unwind it
+        if isinstance(data, list) or isinstance(data, Path):
+            res=[]        
+            for x in data:
+                res.append(self._parse_dict(x))
+            return res
+        
+        # If this is an element then make it a dictionary
+        elif isinstance(data, Vertex) or isinstance(data,Edge) or isinstance(data, VertexProperty) or isinstance(data, Property):
+            data=data.__dict__
+        
+        # If this is a scalar then create a Map with it
+        elif not hasattr(data, "__len__") or isinstance(data, str):
+            data = {0: data}
+
+        for (k, v) in data.items():
+            # If the key is a Vertex or an Edge do special processing
+            if isinstance(k, Vertex) or isinstance(k, Edge):
+                k = k.id
+
+            # If the value is a list do special processing to make it a scalar if the list is of length 1
+            if isinstance(v, list) and len(v) == 1:
+                d[k] = v[0]
+            else:
+                d[k] = v
+
+            # If the value is a Vertex or Edge do special processing
+            if isinstance(d[k], Vertex) or isinstance(d[k], Edge) or isinstance(d[k], VertexProperty) or isinstance(d[k], Property):
+                d[k] = d[k].__dict__
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Modifying `object.__dict__` directly or writing to an instance of a class `__dict__` attribute directly is not recommended. Inside every module is a `__dict__` [object.dict](https://docs.python.org/3/stdtypes.html#object.dict) attribute which contains its symbol table. If you modify `object.__dict__`, then the symbol table is changed. Also, direct assignment to the `__dict__` attribute is not possible.

[Learn more](https://docs.python.org/3/library/stdtypes.html#object.__dict__)